### PR TITLE
Revise revisions

### LINKED ISSUE #31: Support revisions on global commands
From what I could tell, revisions are currently only supported on error patterns. Would be nice to also support them on global commands, so that one can do things like
```
//@revisions: isolated free
//@only-target-linux
//[free]@compile-flags: -Zmiri-disable-isolation
```

FWIW I would have probably picked this syntax
```
//@revisions: isolated free
//@only-target-linux
//@[free]compile-flags: -Zmiri-disable-isolation
```
but that would be inconsistent with `//[fn_ptr]~^ ERROR: pattern`... maybe we should change that to `//~[fn_ptr]^ ERROR: pattern`, too, so that the first character after the `//` always indicates what this is?

OTOH, that might create parsing ambiguities if there is no ERROR level, and `//[revision]` is also consistent, so... whatever.

### PR DESCRIPTION
**Pull request by oli-obk**
fixes #31

### CODE REVIEW

File: src/parser.rs [hunk 1]
```diff
@@ -21,6 +21,40 @@ pub(crate) struct Comments {
     pub ignore: Vec<Condition>,
     /// Only run this test if all of these filters apply
     pub only: Vec<Condition>,
```
Comment (RalfJung) [inline src/parser.rs:23]: There's no fundamental reason these couldn't also be revisioned, is there?

File: src/parser.rs [hunk 2]
```diff
@@ -21,6 +21,40 @@ pub(crate) struct Comments {
...
+    /// Comments that are only available under specific revisions.
+    /// The defaults are in key `vec![]`
+    pub revisioned: HashMap<Vec<String>, Revisioned>,
+}
+
+impl Comments {
+    pub fn find_one_for_revision<'a, T: 'a>(
```
Comment (RalfJung) [inline src/parser.rs:30]: This needs a doc comment. :)

File: src/parser.rs (hunk 1 above)
Comment (oli-obk) [reply src/parser.rs:23]: Yea, I thought for a moment they couldn't be checked, but I think we have all the information available

File: src/lib.rs [hunk 3]
```diff
@@ -715,18 +726,18 @@ fn check_annotations(
...
     }
 
-    let filter = |msgs: Vec<Message>| -> Vec<_> {
+    let filter = |msgs: Vec<Message>, errors: &mut Vec<_>| -> Vec<_> {
         msgs.into_iter()
             .filter(|msg| {
                 msg.level
                     >= comments
-                        .require_annotations_for_level
+                        .find_one_for_revision(revision, |r| r.require_annotations_for_level, |_| errors.push(Error::InvalidComment { msg: "`require_annotations_for_level` specified twice for same revision".into(), line: 0 }))
```
Comment (RalfJung) [inline src/lib.rs:734]: Can you let-bind the long `comments....` expression here? This is an awfully long line

Comment (RalfJung) [reply src/lib.rs:734]: Also I think this can be moved out of the `filter` loop?

Comment (oli-obk) [reply src/lib.rs:734]: oh wow rustfmt just completely gave up :D yea I'll clean that up

File: src/lib.rs [hunk 4]
```diff
@@ -727,11 +727,21 @@ fn check_annotations(
...
+                msg: "`require_annotations_for_level` specified twice for same revision".into(),
+                line: 0,
+            })
+        };
         msgs.into_iter()
             .filter(|msg| {
                 msg.level
                     >= comments
-                        .find_one_for_revision(revision, |r| r.require_annotations_for_level, |_| errors.push(Error::InvalidComment { msg: "`require_annotations_for_level` specified twice for same revision".into(), line: 0 }))
+                        .find_one_for_revision(
```
Comment (RalfJung) [inline src/lib.rs:740]: Would it make sense to call this outside the `filter` closure? Right now we are re-doing this search over and over.

Comment (oli-obk) [reply src/lib.rs:740]: oh 🤦 I thought we needed the iteration

**Pull request merged**